File: tests/test_npcc.py
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from npcc import PATHWAY, compress, decompress
from npcc.format import unpack


def check(x: bytes) -> bytes:
    c = compress(x)
    y = decompress(c)
    assert y == x, (len(x), len(y), unpack(c)["pathway_id"])
    return c


def test_cases():
    samples = [
        b"",
        b"\x00",
        b"\x00" * 100,
        b"\x00" * 2000,
        b"A" * 500,
        b"ab" * 400,
        b"The quick brown fox jumps over the lazy dog. " * 20,
        bytes(range(256)),
        bytes(range(256)) * 4,
        b"\xff\x00" * 300,
    ]
    for x in samples:
        c = check(x)
        info = unpack(c)
        print(f"{len(x):6d} -> {len(c):6d}  path={PATHWAY(info['pathway_id']).name}")


if __name__ == "__main__":
    test_cases()
    print("ok")

File: tests/test_npcc.c
#include "npcc.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int check(const uint8_t *x, size_t n) {
    uint8_t *c = NULL, *y = NULL;
    size_t cn = 0, yn = 0;
    int rc = npcc_compress(x, n, NULL, &c, &cn);
    if (rc) {
        fprintf(stderr, "compress %zu: %s\n", n, npcc_strerror(rc));
        return 1;
    }
    rc = npcc_decompress(c, cn, &y, &yn);
    if (rc || yn != n || memcmp(y, x, n) != 0) {
        int path = -1;
        npcc_peek(c, cn, &path, NULL);
        fprintf(stderr, "roundtrip fail n=%zu packed=%zu path=%s rc=%s\n", n, cn,
                npcc_pathway_name(path), npcc_strerror(rc));
        free(c);
        free(y);
        return 1;
    }
    int path = -1;
    npcc_peek(c, cn, &path, NULL);
    printf("%6zu -> %6zu  path=%s\n", n, cn, npcc_pathway_name(path));
    free(c);
    free(y);
    return 0;
}

int main(void) {
    int fail = 0;
    fail |= check((const uint8_t *)"", 0);
    uint8_t z1[1] = {0};
    fail |= check(z1, 1);
    uint8_t z100[100];
    memset(z100, 0, 100);
    fail |= check(z100, 100);
    uint8_t *z2000 = calloc(2000, 1);
    fail |= check(z2000, 2000);
    uint8_t a500[500];
    memset(a500, 'A', 500);
    fail |= check(a500, 500);
    uint8_t ab[800];
    for (int i = 0; i < 800; i++) ab[i] = (uint8_t)("ab"[i & 1]);
    fail |= check(ab, 800);
    const char *fox = "The quick brown fox jumps over the lazy dog. ";
    uint8_t foxr[900];
    size_t fl = strlen(fox);
    for (int i = 0; i < 20; i++) memcpy(foxr + i * fl, fox, fl);
    fail |= check(foxr, 20 * fl);
    uint8_t r256[256];
    for (int i = 0; i < 256; i++) r256[i] = (uint8_t)i;
    fail |= check(r256, 256);
    uint8_t r1024[1024];
    for (int i = 0; i < 1024; i++) r1024[i] = (uint8_t)(i & 255);
    fail |= check(r1024, 1024);
    uint8_t ff00[600];
    for (int i = 0; i < 600; i++) ff00[i] = (uint8_t)((i & 1) ? 0 : 0xff);
    fail |= check(ff00, 600);
    free(z2000);
    if (fail) return 1;
    puts("ok");
    return 0;
}
